### ISSUE TITLE
recommended hardware resources

### ISSUE THREAD

**Issue by aswanthkrishna** [AIoT-MLSys-Lab/SVD-LLM#3]
What is the minimum hardware resources required to test out this codebase for llama 7B.

**Comment by tuidan** [comment]
Hi! We strongly recommend to compress llama 7B on A100. However, it is still possible to run the compression with only 20G GPU memory by the following command:

`python SVDLLM.py 
--step 1  
--run_low_resource
--ratio COMPRESSION_RATIO 
--model HUGGINGFACE_MODEL_REPO 
--whitening_nsamples WHITENING_SAMPLE_NUMBER 
--dataset WHITENING_DATASET 
--seed SAMPLING_SEED 
--model_seq_len MODEL_SEQ_LEN 
--save_path WHITENING_INFO_SAVING_PATH`

**Comment by tuidan** [comment]
> Hi @tuidan,
> 
> I've been running the following command on an A100 80GB GPU with ~90GB of CPU RAM:
> 
> ```
> python SVDLLM.py --model jeffwan/llama-7b-hf --step 1 --ratio 0.2 --whitening_nsamples 256 --dataset wikitext2 --seed 3 --model_seq_len 2048 --save_path . --run_low_resource
> ```
> 
> However, the program consistently gets suspended due to CPU RAM shortages. It seems the issue might be related to the loop iterating through layers in this part of the code:
> 
> https://github.com/AIoT-MLSys-Lab/SVD-LLM/blob/7dc65bd8504590ecd594a096db09c9b1bbc84ab0/SVDLLM.py#L136
> 
> 
> Could you share the machine configuration you used to run this? Also, do you have any further recommendations for resolving this issue?
> Thanks in advance for your help!

In fact, our code requires at least 100GB CPU RAM since it needs to profile and cache the whitening matrices of all the weight matrices before running the compression. The reason to do this is because recomputing the whitening matrix will take about 10-15 min. If we cache it, we only need about less than 5 min to run the compression. 

- Considering that your CPU RAM is about 90GB, I recommend you to transform the whitenining matrix to fp16 before caching it, i.e. modify the code in this [line](https://github.com/AIoT-MLSys-Lab/SVD-LLM/blob/7dc65bd8504590ecd594a096db09c9b1bbc84ab0/SVDLLM.py#L72) to `layer_profile[name] = scaling_diag_matrix.half().cpu()` .

- If your CPU memory is less than 60GB, I recommend you to directly compute the whitening matrix and then run the compression weight by weight rather than caching all the whitening matrices beforehand.

**Comment by hsb1995** [comment]
> > Hi [@tuidan](https://github.com/tuidan),
> > I've been running the following command on an A100 80GB GPU with ~90GB of CPU RAM:
> > ```
> > python SVDLLM.py --model jeffwan/llama-7b-hf --step 1 --ratio 0.2 --whitening_nsamples 256 --dataset wikitext2 --seed 3 --model_seq_len 2048 --save_path . --run_low_resource
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, the program consistently gets suspended due to CPU RAM shortages. It seems the issue might be related to the loop iterating through layers in this part of the code:
> > 
> >   
> >     
> >       [SVD-LLM/SVDLLM.py](https://github.com/AIoT-MLSys-Lab/SVD-LLM/blob/7dc65bd8504590ecd594a096db09c9b1bbc84ab0/SVDLLM.py#L136)
> >     
> >     
> >          Line 136
> >       in
> >       [7dc65bd](/AIoT-MLSys-Lab/SVD-LLM/commit/7dc65bd8504590ecd594a096db09c9b1bbc84ab0)
> >     
> >   
> >   
> >     
> > 
> >         
> >           
> >            for i in tqdm(range(len(layers))): 
> >         
> >     
> >   
> > 
> > 
> > Could you share the machine configuration you used to run this? Also, do you have any further recommendations for resolving this issue?
> > Thanks in advance for your help!
> 
> In fact, our code requires at least 100GB CPU RAM since it needs to profile and cache the whitening matrices of all the weight matrices before running the compression. The reason to do this is because recomputing the whitening matrix will take about 10-15 min. If we cache it, we only need about less than 5 min to run the compression.
> 
> * Considering that your CPU RAM is about 90GB, I recommend you to transform the whitenining matrix to fp16 before caching it, i.e. modify the code in this [line](https://github.com/AIoT-MLSys-Lab/SVD-LLM/blob/7dc65bd8504590ecd594a096db09c9b1bbc84ab0/SVDLLM.py#L72) to `layer_profile[name] = scaling_diag_matrix.half().cpu()` .
> * If your CPU memory is less than 60GB, I recommend you to directly compute the whitening matrix and then run the compression weight by weight rather than caching all the whitening matrices beforehand.

If your CPU memory is less than 60GB, I recommend you to directly compute the whitening matrix and then run the compression weight by weight rather than caching all the whitening matrices beforehand.

Dear professor, how can you fully utilize your code if it is less than 60GB. I really think your code is very good and I want to learn how to run it smoothly

**Comment by hsb1995** [comment]
import torch
import gc

# tqdm 进度条
for i in tqdm(range(len(layers))):
    layer_profile = {}  # 只存当前层的信息
    layer = layers[i].to(dev)  # 将当前层移动到设备

    # 找到所有子层
    subset = find_layers(layer)

    # 定义 hook 函数，在前向传播时计算 scaling_diag_matrix
    def hook(module, input, output):
        inp = input[0].detach().float()
        if inp.dim() == 2:  # 处理 OPT 模型
            inp = inp.unsqueeze(0)
        adds = torch.matmul(inp.transpose(1,2), inp)
        adds_sum = torch.sum(adds, dim=0)
        module.scaling_diag_matrix += adds_sum
        del inp, adds, adds_sum, output
        torch.cuda.empty_cache()
        gc.collect()

    # 注册 hook
    handles = []
    for name in subset:
        subset[name].scaling_diag_matrix = torch.zeros(
            (subset[name].weight.shape[0], subset[name].weight.shape[0]),
            dtype=torch.float32, device=dev  # 直接在 GPU 上创建，减少 CPU->GPU 传输
        )
        handles.append(subset[name].register_forward_hook(hook))

    # 逐个计算前向传播，避免一次性加载过多数据
    for j in range(inps.shape[0]):
        if "opt" not in model_name:
            outs[j] = layer(
                inps[j].unsqueeze(0),
                attention_mask=attention_masks[j].unsqueeze(0).to(dev),
                position_ids=position_ids[j].unsqueeze(0).to(dev)
            )[0]
        else:
            outs[j] = layer(
                inps[j].unsqueeze(0),
                attention_mask=attention_masks[j].unsqueeze(0).to(dev)
            )[0]

    # 移除 hook，减少显存占用
    for h in handles:
        h.remove()
    del handles
    torch.cuda.empty_cache()
    gc.collect()

    # 计算 whitening matrix，并直接用它进行 SVD 计算（不缓存）
    for name in subset:
        raw_scaling_diag_matrix = subset[name].scaling_diag_matrix.double().to(dev)

        # Cholesky 分解，修正非正定矩阵
        try:
            scaling_diag_matrix = torch.linalg.cholesky(raw_scaling_diag_matrix)
        except Exception:
            print("Warning: eigen scaling_diag_matrix is not positive!")
            eigenvalues = torch.linalg.eigvalsh(raw_scaling_diag_matrix)
            raw_scaling_diag_matrix += (-eigenvalues[0] + 1e-6) * torch.eye(
                raw_scaling_diag_matrix.shape[0], device=dev
            )
            scaling_diag_matrix = torch.linalg.cholesky(raw_scaling_diag_matrix)
            del eigenvalues

        # **不缓存所有 whitening matrices**
        layer_profile[name] = scaling_diag_matrix.cpu()

        # 释放内存
        del scaling_diag_matrix, raw_scaling_diag_matrix
        subset[name].scaling_diag_matrix = None
        torch.cuda.empty_cache()
        gc.collect()

    # **逐层处理后，立刻释放 layer**
    layers[i] = layer.cpu()
    profiling_mat[i] = layer_profile
    inps = outs  # 更新输入
    del layer, subset, layer_profile
    torch.cuda.empty_cache()
    gc.collect()

# 返回 profiling 结果
return profiling_mat



Dear professor，can i revise this ok?
